File: scratch/check_brackets.dart
import 'dart:io';

void main() {
  final file = File(r"c:\Users\SamDa\arion_daq\lib\views\analysis_tab.dart");
  final content = file.readAsStringSync();
  
  final stack = <Map<String, dynamic>>[];
  final lines = content.split('\n');
  
  for (var i = 0; i < lines.length; i++) {
    final line = lines[i];
    for (var j = 0; j < line.length; j++) {
      final char = line[j];
      if (char == '(' || char == '{' || char == '[') {
        stack.add({
          'char': char,
          'line': i + 1,
          'col': j + 1,
        });
      } else if (char == ')' || char == '}' || char == ']') {
        if (stack.isEmpty) {
          print("Extra closing $char at line ${i+1}, col ${j+1}");
          continue;
        }
        final top = stack.removeLast();
        final topChar = top['char'];
        if ((char == ')' && topChar != '(') || 
            (char == '}' && topChar != '{') || 
            (char == ']' && topChar != '[')) {
          print("Mismatched $char at line ${i+1}, col ${j+1} (opened with $topChar at line ${top['line']}, col ${top['col']})");
        }
      }
    }
  }
  
  if (stack.isNotEmpty) {
    print("Unclosed brackets:");
    for (var item in stack) {
      print("  ${item['char']} at line ${item['line']}, col ${item['col']}");
    }
  } else {
    print("All brackets match perfectly!");
  }
}

File: scratch/check_brackets.py
def check_brackets(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    
    stack = []
    lines = content.split('\n')
    for i, line in enumerate(lines):
        for j, char in enumerate(line):
            if char in '({[':
                stack.append((char, i+1, j+1))
            elif char in ')}]':
                if not stack:
                    print(f"Extra closing {char} at line {i+1}, col {j+1}")
                    continue
                top, line_num, col_num = stack.pop()
                if (char == ')' and top != '(') or (char == '}' and top != '{') or (char == ']' and top != '['):
                    print(f"Mismatched {char} at line {i+1}, col {j+1} (opened with {top} at line {line_num}, col {col_num})")
                    
    if stack:
        print("Unclosed brackets:")
        for top, line_num, col_num in stack:
            print(f"  {top} at line {line_num}, col {col_num}")

check_brackets(r"c:\Users\SamDa\arion_daq\lib\widgets\zoomable_chart.dart")
